### PR TITLE
Fix rebin

### PR DESCRIPTION
**Pull request by iprafols**
This PR updates the function rebin. There are two changes. 
First, cuts in the wavelength (lambda_min, lambda_max, lambda_rest_frame_min, and lambda_rest_frame_max) are moved prior to the change in baseline wavelength. This makes more sense since we are using the actual measured values to limit the wavelength range. 

The second change is necessary as there are some instabilities in the way the binning is computed. A new function is added that does the mapping from original to common binning in a more stable way that does not require the pixels to be homogeneously distributed. This requires the definition of log_lambda_grid and similar variables which make Forest.log_lambda_max (and similar) useless. I have also removed the relevant references from other places

### CODE REVIEW

**Review by Waelthus — COMMENTED** [review]
The new approach in general seems like a good idea!
Some comments about specifics added.
I'll run a test analysis for pk1d (on a merge of this with #850 ) to see how/if this changes/breaks things there.

The main things I'd like to be added is a choice for the size of restframe pixels and a check of grid alignment with the actual grid in the data (if this is not there yet).

File: py/picca/delta_extraction/utils.py [hunk 1]
```diff
@@ -126,6 +127,10 @@ def class_from_string(class_name, module_name):
         accepted_options = []
     return class_object, default_args, accepted_options
 
+def find_bins(original_array, grid_array):
+    """For each element in original_array, find the corresponding bin in grid_array"""
+    found_bin = (np.abs(grid_array - original_array[:,None])).argmin(axis=1)
```
[the remark this answers is not in the document]
Comment (iprafols) [reply py/picca/delta_extraction/utils.py]: Done, but do check the syntax is ok (I did not know this existed)

File: py/picca/delta_extraction/astronomical_objects/forest.py [hunk 2]
```diff
@@ -584,3 +534,55 @@ def rebin(self):
...
+                pixel_step)
+            cls.mask_fields = defaults.get("mask fields log").copy()
+        elif wave_solution == "lin":
+            cls.lambda_grid = np.arange(
+                lambda_min,
+                lambda_max+ pixel_step/2,
+                pixel_step)
+            cls.lambda_rest_frame_grid = np.arange(
+                lambda_min_rest_frame,
+                lambda_max_rest_frame + pixel_step/2,
```
[the remark this answers is not in the document]
Comment (iprafols) [reply py/picca/delta_extraction/astronomical_objects/forest.py:579]: On the separate pixel size, I think we need this, but this is more for another PR (probably the Pk1d one), here I'm just improving the stability of the rebin method (thus leaving the other things as they were.

File: py/picca/delta_extraction/data.py [hunk 3]
```diff
@@ -7,18 +7,27 @@
...
 
 defaults = {
     "analysis type": "BAO 3D",
     "lambda abs IGM": "LYA",
+    "lambda max": 5500.0,
+    "lambda max rest frame": 1200.0,
+    "lambda min": 3600.0,
+    "lambda min rest frame": 1040.0,
     "minimum number pixels in forest": 50,
+    "rebin": 3,
```
[the remark this answers is not in the document]
Comment (iprafols) [reply py/picca/delta_extraction/data.py:30]: Now that you mention this, I am more inclined to simply remove this option (and have the user pass the delta_log_lambda of their choice. We can maintain this option in SdssData and modify the `delta_log_lambda` parameter accordingly. I'll do a commit to fix this

[the remark this answers is not in the document]
Comment (iprafols) [reply py/picca/delta_extraction/data.py]: pushed

File: py/picca/delta_extraction/astronomical_objects/forest.py [hunk 4]
```diff
@@ -584,3 +534,55 @@ def rebin(self):
...
+                np.log10(lambda_min),
+                np.log10(lambda_max)+ pixel_step/2,
+                pixel_step)
+            cls.log_lambda_rest_frame_grid = np.arange(
+                np.log10(lambda_min_rest_frame),
+                np.log10(lambda_max_rest_frame)+ pixel_step/2,
+                pixel_step)
+            cls.mask_fields = defaults.get("mask fields log").copy()
+        elif wave_solution == "lin":
+            cls.lambda_grid = np.arange(
```
[the remark this answers is not in the document]
Comment (iprafols) [reply py/picca/delta_extraction/astronomical_objects/forest.py:573]: We probably should, but we are not doing in on master either, so I think we should file an issue and address it on another PR

File: py/picca/delta_extraction/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/delta_extraction/utils.py]: looks good

**Review by Waelthus — APPROVED** [review]
looks good, added a comma to fix a syntax error. runtime tests running now.

File: py/picca/delta_extraction/data_catalogues/desi_healpix.py [hunk 5]
```diff
@@ -234,7 +234,45 @@ def read_file(self, filename, catalogue, forests_by_targetid):
...
+                            f.write(f"{log_lambda} {flux}\n")
+                        f.close()
+                    else:
+                        f = open(f"{self.out_dir}rebin_files/forest_flux_before_rebin_linbin_losid{forest.los_id}_spec{spec_name}.txt", "w")
+                        f.write("# wavelength flux\n")
+                        for lambda_, flux in zip(forest.lambda_, forest.flux):
+                            f.write(f"{lambda_} {flux}\n")
+                        f.close()
                 forest.rebin()
+                if forest.los_id in [39633561183060922, 39633485232605672,
```
[the remark this answers is not in the document]
Comment (iprafols) [reply py/picca/delta_extraction/data_catalogues/desi_healpix.py]: yes, sorry leftover from tests.

**Pull request merged**